R/as.data.table.R

```diff
@@ -241,17 +242,22 @@ as.data.table.data.frame = function(x, keep.rownames=FALSE, key=NULL, ...) {
   # fix for #1078 and #1128, see .resetclass() for explanation.
   setattr(ans, "class", .resetclass(x, "data.frame"))
   setalloccol(ans)
-  setkeyv(ans, key)
+  if (!is.null(key)) setkeyv(ans, key)
   ans
 }
 
-as.data.table.data.table = function(x, ...) {
+as.data.table.data.table = function(x, keep.rownames = FALSE, key = NULL, ...) {
   # as.data.table always returns a copy, automatically takes care of #473
   if (any(cols_with_dims(x))) { # for test 2089.2
-    return(as.data.table.list(x, ...))
+    return(as.data.table.list(x, keep.rownames = keep.rownames, key = key, ...))
   }
   x = copy(x) # #1681
   # fix for #1078 and #1128, see .resetclass() for explanation.
   setattr(x, 'class', .resetclass(x, "data.table"))
+  if (!is.null(key)) {
+    setkeyv(x, key)
+  } else {
+    setkeyv(x, NULL)
+  }
```

Comment:
```suggestion
  if (!missing(key)) setkeyv(x, key)
```
is much simpler and should have the same behavior